## Flight Price Prediction

### 1) Problem statement
- This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.


### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
- The data consists of 8 column and 1000 rows.

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [139]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

#### Import the CSV Data as Pandas DataFrame

In [140]:
df = pd.read_excel('data/IndianFlightdata.xlsx')

#### Show Top 5 Records

In [141]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

#### Shape of the dataset

In [142]:
df.shape

(10683, 11)

### 2.2 Dataset information

The various features of the dataset are explained below:
1) Airline: The name of the airline company is stored in the airline column. It is a categorical feature having 6 different airlines.
2) Flight: Flight stores information regarding the plane's flight code. It is a categorical feature.
3) Source City: City from which the flight takes off. It is a categorical feature having 6 unique cities.
4) Departure Time: This is a derived categorical feature obtained created by grouping time periods into bins. It stores information about the departure time and have 6 unique time labels.
5) Stops: A categorical feature with 3 distinct values that stores the number of stops between the source and destination cities.
6) Arrival Time: This is a derived categorical feature created by grouping time intervals into bins. It has six distinct time labels and keeps information about the arrival time.
7) Destination City: City where the flight will land. It is a categorical feature having 6 unique cities.
8) Class: A categorical feature that contains information on seat class; it has two distinct values: Business and Economy.
9) Duration: A continuous feature that displays the overall amount of time it takes to travel between cities in hours.
10)Days Left: This is a derived characteristic that is calculated by subtracting the trip date by the booking date.
11) Price: Target variable stores information of the ticket price.

### 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

### 3.1 Check Missing values

In [143]:
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

##### There are 2 missing values in the dataset

### 3.2 Check Duplicates

In [144]:
df.duplicated().sum()

np.int64(220)

##### There are 220 duplicated values in the dataset

### 3.3 Check data types

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


### 3.4 Checking the number of unique values of each column

In [146]:
df.nunique()

Airline              12
Date_of_Journey      44
Source                5
Destination           6
Route               128
Dep_Time            222
Arrival_Time       1343
Duration            368
Total_Stops           5
Additional_Info      10
Price              1870
dtype: int64

### 3.5 Check statistics of data set

#### Insights from EDA (so far)
1. The dataset is mostly complete, with only 2 missing values and 220 duplicate rows detected. This suggests good data quality but highlights the need to address duplicates.
2. The data contains a mix of categorical (e.g., airline, cities, class) and continuous (e.g., duration, days left, price) features, with several columns having a limited set of unique values suitable for analysis.
3. Initial statistics and info checks provide a clear overview of the data's structure and distribution, setting the stage for further exploration and feature engineering.

In [147]:
df.describe()

Price
count  10683.000000
mean    9087.064121
std     4611.359167
min     1759.000000
25%     5277.000000
50%     8372.000000
75%    12373.000000
max    79512.000000

#### Insights from EDA (so far)
1. The dataset is mostly complete, with only 2 missing values and 220 duplicate rows detected.
2. The data contains a mix of categorical (e.g., airline, cities, class) and continuous (e.g., duration, days left, price) features, with several columns having a limited set of unique values suitable for analysis.

### 3.6 Data Cleaning

##### 3.6.1 Remove Duplicates

In [148]:
df = df.drop_duplicates()

##### 3.6.2 Handle missing values

In [149]:
df = df.dropna()

##### 3.6.3 Removing the outliers

In [150]:
Q1 = df['Price'].quantile(0.25)
Q3 = df['Price'].quantile(0.75)
IQR = Q3 - Q1
df = df[~((df['Price'] < (Q1 - 1.5 * IQR)) | (df['Price'] > (Q3 + 1.5 * IQR)))]

In [151]:
df.reset_index(drop=True, inplace=True)

#### Data Cleaning Insights

1. Duplicates and missing values have been removed.
2. Outliers in the target variable ('Price') have been filtered out using the IQR method.
3. The dataset index has been reset.

### 3.7 Exploring Data

In [152]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

### 3.7.1 Feature Engineering

#### Creating columns from Date_of_Journey Column

In [153]:
# Creating separate columns Date, Month, Year from Date_of_Journey column
df['Date'] = df['Date_of_Journey'].str.split('/').str[0].astype(int)
df['Month'] = df['Date_of_Journey'].str.split('/').str[1].astype(int)
df['Year'] = df['Date_of_Journey'].str.split('/').str[2].astype(int)

#### Creating columns from Arrival_Time Column

In [154]:
# Creating columns Arrival_Hour and Arrival_Minute from Arrival column
df['Arrival_Time'] = df['Arrival_Time'].apply(lambda x:x.split(' ')[0])
df['Arrival_Hour'] = df['Arrival_Time'].str.split(':').str[0].astype(int)
df['Arrival_Minute'] = df['Arrival_Time'].str.split(':').str[1].astype(int)

#### Creating columns from Dep_Time Column

In [155]:
# Creating columns Departure_Hour and Departure_Minute from Dep_Time column
df['Departure_Hour'] = df['Dep_Time'].str.split(':').str[0].astype(int)
df['Departure_Minute'] = df['Dep_Time'].str.split(':').str[1].astype(int)

#### Mapping Total_Stops column

In [156]:
df['Total_Stops'] = df['Total_Stops'].map({'non-stop':0, '1 stop':1, '2 stops':2, '3 stops':3, '4 stops':4})

#### Removing unwanted columns

In [157]:
# Removing unwanted columns
df.drop(['Date_of_Journey', 'Arrival_Time', 'Dep_Time', 'Duration','Route','Additional_Info'], axis=1, inplace=True)

#### Saving the file for Data Ingestion

In [171]:
import os

save_path = os.path.join("..", "Flight_Data", "flight_data.csv")
df.to_csv(save_path, index=False)

### 4. Exploring Data ( Visualization )

In [159]:
df.head()

Airline    Source Destination  Total_Stops  Price  Date  Month  Year  \
0       IndiGo  Banglore   New Delhi            0   3897    24      3  2019   
1    Air India   Kolkata    Banglore            2   7662     1      5  2019   
2  Jet Airways     Delhi      Cochin            2  13882     9      6  2019   
3       IndiGo   Kolkata    Banglore            1   6218    12      5  2019   
4       IndiGo  Banglore   New Delhi            1  13302     1      3  2019   

   Arrival_Hour  Arrival_Minute  Departure_Hour  Departure_Minute  
0             1              10              22                20  
1            13              15               5                50  
2             4              25               9                25  
3            23              30              18                 5  
4            21              35              16                50